src/main/java/com/iota/iri/service/tipselection/impl/WalkValidatorImpl.java

```diff
@@ -75,19 +82,30 @@ public boolean isValid(Hash transactionHash) throws Exception {
         return true;
     }
 
-    private boolean belowMaxDepth(Hash tip, int depth) throws Exception {
+    private boolean belowMaxDepth(Hash tip, int allowedSnapshotIndex) throws Exception {
         //if tip is confirmed stop
-        if (TransactionViewModel.fromHash(tangle, tip).snapshotIndex() >= depth) {
+        if (TransactionViewModel.fromHash(tangle, tip).snapshotIndex() >= allowedSnapshotIndex) {
             return false;
         }
         //if tip unconfirmed, check if any referenced tx is confirmed below maxDepth
         Queue<Hash> nonAnalyzedTransactions = new LinkedList<>(Collections.singleton(tip));
         Set<Hash> analyzedTransactions = new HashSet<>();
         Hash hash;
         while ((hash = nonAnalyzedTransactions.poll()) != null) {
+            if (FAILED_BELOW_MAX_DEPTH_CACHE.contains(hash)) {
+                updateCache(analyzedTransactions);
+                return true;
```

Comment:
if it `failed` before (I'm assuming that that is what the cahce is for) shouldn't this return `false`?

Reply:
Could you add a javadoc comment to this method as you're touching it?

Reply:
BelowMaxDepth is a negative predicate (returning true means its "bad", failing is also bad hence true) - this is annoying to reason about and can lead to bugs. I think we should invert it. Maybe in another PR.